### ISSUE TITLE
tsMuxeR doesn't start on macOS

### ISSUE THREAD

**Issue by onon765trb** [UniversalMediaServer/UniversalMediaServer#1176]
It isn't impossible to playback movie wuth tsmuxer.

[debug.log.zip](https://github.com/UniversalMediaServer/UniversalMediaServer/files/757134/debug.log.zip)

**Comment by onon765trb** [comment]
I recommend the usage of `  -fflags genpts` when using `-codec copy`.
This is because there is when not be able to copy.

For example
Without   `  -fflags genpts`
/Applications/Universal\ Media\ Server_v8.app/Contents/Resources/osx/ffmpeg -i test.wmv  -c:v copy test.ts
ffmpeg version N-80408-g39c0b22-tessus Copyright (c) 2000-2016 the FFmpeg developers
  built with Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
  configuration: --cc=/usr/bin/clang --prefix=/opt/ffmpeg --as=yasm --extra-version=tessus --enable-avisynth --enable-fontconfig --enable-gpl --enable-libass --enable-libbluray --enable-libfreetype --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopus --enable-libschroedinger --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-libzmq --enable-version3 --disable-ffplay --disable-indev=qtkit --disable-indev=x11grab_xcb
  libavutil      55. 24.100 / 55. 24.100
  libavcodec     57. 44.100 / 57. 44.100
  libavformat    57. 37.101 / 57. 37.101
  libavdevice    57.  0.101 / 57.  0.101
  libavfilter     6. 46.100 /  6. 46.100
  libswscale      4.  1.100 /  4.  1.100
  libswresample   2.  0.101 /  2.  0.101
  libpostproc    54.  0.100 / 54.  0.100
Input #0, asf, from test.wmv:
  Duration: 01:08:00.05, start: 0.000000, bitrate: 6014 kb/s
    Stream #0:0(jpn): Audio: wmapro (b[1][0][0] / 0x0162), 44100 Hz, stereo, fltp, 256 kb/s
    Stream #0:1(jpn): Video: vc1 (Advanced) (WVC1 / 0x31435657), yuv420p, 1920x1080, 5744 kb/s, 30 fps, 30 tbr, 1k tbn, 1k tbc
[mpegts @ 0x7fd30b0df200] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
    Last message repeated 1 times
Output #0, mpegts, to 'test.ts':
  Metadata:
    encoder         : Lavf57.37.101
    Stream #0:0(jpn): Video: vc1 (WVC1 / 0x31435657), yuv420p, 1920x1080, q=2-31, 5744 kb/s, 30 fps, 30 tbr, 90k tbn, 1k tbc
    Stream #0:1(jpn): Audio: mp2, 44100 Hz, stereo, s16, 384 kb/s
    Metadata:
      encoder         : Lavc57.44.100 mp2
Stream mapping:
  Stream #0:1 -> #0:0 (copy)
  Stream #0:0 -> #0:1 (wmapro (native) -> mp2 (native))
Press [q] to stop, [?] for help
[mpegts @ 0x7fd30b0df200] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
[mpegts @ 0x7fd30b0df200] first pts value must be set
av_interleaved_write_frame(): Invalid data found when processing input
[mpegts @ 0x7fd30b0df200] first pts value must be set
Error writing trailer of test.ts: Invalid data found when processing inputframe=   96 fps=0.0 q=-1.0 Lsize=       3kB time=00:00:03.16 bitrate=   8.1kbits/s speed= 643x    
video:2206kB audio:2kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
Conversion failed!

With   `  -fflags genpts`
/Applications/Universal\ Media\ Server_v8.app/Contents/Resources/osx/ffmpeg  -fflags genpts  -i test.wmv  -c:v copy test.ts
ffmpeg version N-80408-g39c0b22-tessus Copyright (c) 2000-2016 the FFmpeg developers
  built with Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
  configuration: --cc=/usr/bin/clang --prefix=/opt/ffmpeg --as=yasm --extra-version=tessus --enable-avisynth --enable-fontconfig --enable-gpl --enable-libass --enable-libbluray --enable-libfreetype --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopus --enable-libschroedinger --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-libzmq --enable-version3 --disable-ffplay --disable-indev=qtkit --disable-indev=x11grab_xcb
  libavutil      55. 24.100 / 55. 24.100
  libavcodec     57. 44.100 / 57. 44.100
  libavformat    57. 37.101 / 57. 37.101
  libavdevice    57.  0.101 / 57.  0.101
  libavfilter     6. 46.100 /  6. 46.100
  libswscale      4.  1.100 /  4.  1.100
  libswresample   2.  0.101 /  2.  0.101
  libpostproc    54.  0.100 / 54.  0.100
Input #0, asf, from test.wmv:
  Duration: 01:08:00.05, start: 0.000000, bitrate: 6014 kb/s
    Stream #0:0(jpn): Audio: wmapro (b[1][0][0] / 0x0162), 44100 Hz, stereo, fltp, 256 kb/s
    Stream #0:1(jpn): Video: vc1 (Advanced) (WVC1 / 0x31435657), yuv420p, 1920x1080, 5744 kb/s, 30 fps, 30 tbr, 1k tbn, 1k tbc
[mpegts @ 0x7fdfaf0df200] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
    Last message repeated 1 times
Output #0, mpegts, to 'test.ts':
  Metadata:
    encoder         : Lavf57.37.101
    Stream #0:0(jpn): Video: vc1 (WVC1 / 0x31435657), yuv420p, 1920x1080, q=2-31, 5744 kb/s, 30 fps, 30 tbr, 90k tbn, 1k tbc
    Stream #0:1(jpn): Audio: mp2, 44100 Hz, stereo, s16, 384 kb/s
    Metadata:
      encoder         : Lavc57.44.100 mp2
Stream mapping:
  Stream #0:1 -> #0:0 (copy)
  Stream #0:0 -> #0:1 (wmapro (native) -> mp2 (native))
Press [q] to stop, [?] for help
[mpegts @ 0x7fdfaf0df200] Non-monotonous DTS in output stream 0:1; previous: 5252412, current: 5250539; changing to 5252413. This may result in incorrect timestamps in the output file.
[mpegts @ 0x7fdfaf0df200] Non-monotonous DTS in output stream 0:1; previous: 125873924, current: 125872100; changing to 125873925. This may result in incorrect timestamps in the output file.
[mpegts @ 0x7fdfaf0df200] Non-monotonous DTS in output stream 0:1; previous: 215802782, current: 215799894; changing to 215802783. This may result in incorrect timestamps in the output file.
[mpegts @ 0x7fdfaf0df200] Non-monotonous DTS in output stream 0:1; previous: 215802783, current: 215802245; changing to 215802784. This may result in incorrect timestamps in the output file.
[mpegts @ 0x7fdfaf0df200] Non-monotonous DTS in output stream 0:1; previous: 234843663, current: 234841869; changing to 234843664. This may result in incorrect timestamps in the output file.
[mpegts @ 0x7fdfaf0df200] Non-monotonous DTS in output stream 0:1; previous: 292547135, current: 292545278; changing to 292547136. This may result in incorrect timestamps in the output file.
frame=122400 fps=1326 q=-1.0 Lsize= 3301946kB time=01:08:00.11 bitrate=6629.6kbits/s speed=44.2x    
video:2853308kB audio:191252kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 8.453937%

**Comment by Nadahar** [comment]
@onon765trb 

> This is because there is when not be able to copy.

I simply don't understand this sentence, can you rephrase?

**Comment by Sami32** [comment]
That's interesting:
> DEBUG 2017-02-07 18:08:41.523 [UPNP-AliveMessageSender] Sending ALIVE...
> DEBUG 2017-02-07 18:08:41.593 [UPNP-AliveMessageSender] Error sending ALIVE message
> java.io.IOException: No route to host
> 	at java.net.PlainDatagramSocketImpl.send(Native Method)
> 	at java.net.DatagramSocket.send(DatagramSocket.java:693)
> 	at net.pms.network.UPNPHelper.sendMessage(UPNPHelper.java:386)
> 	at net.pms.network.UPNPHelper.sendMessage(UPNPHelper.java:358)
> 	at net.pms.network.UPNPHelper.sendAlive(UPNPHelper.java:223)
> 	at net.pms.network.UPNPHelper.lambda$listen$0(UPNPHelper.java:409)
> 	at java.lang.Thread.run(Thread.java:745)
> TRACE 2017-02-07 18:08:42.400 [UPNPHelper] Receiving a NOTIFY from [192.168.10.1:47557]
> DEBUG 2017-02-07 18:08:42.525 [cling-10] Ignoring device: InternetGatewayDevice (RemoteDevice) Identity: (RemoteDeviceIdentity) UDN: uuid:09182736-4554-6372-8191-c025a28010ac, Descriptor: http://192.168.10.1:52869/43a59e11.xml, Root: true
> DEBUG 2017-02-07 18:08:42.525 [cling-10] Ignoring embedded device: urn:schemas-upnp-org:device:WANDevice:1 (RemoteDevice) Identity: (RemoteDeviceIdentity) UDN: uuid:09182736-4554-6372-8191-c025a28010ac, Descriptor: http://192.168.10.1:52869/43a59e11.xml, Root: false
> DEBUG 2017-02-07 18:08:51.598 [UPNP-AliveMessageSender] Sending ALIVE...
> DEBUG 2017-02-07 18:08:51.637 [UPNP-AliveMessageSender] Error sending ALIVE message
> java.io.IOException: No route to host
> 	at java.net.PlainDatagramSocketImpl.send(Native Method)
> 	at java.net.DatagramSocket.send(DatagramSocket.java:693)
> 	at net.pms.network.UPNPHelper.sendMessage(UPNPHelper.java:386)
> 	at net.pms.network.UPNPHelper.sendMessage(UPNPHelper.java:358)
> 	at net.pms.network.UPNPHelper.sendAlive(UPNPHelper.java:223)
> 	at net.pms.network.UPNPHelper.lambda$listen$0(UPNPHelper.java:409)
> 	at java.lang.Thread.run(Thread.java:745)
> TRACE 2017-02-07 18:08:56.021 [UPNPHelper] Access granted to /xxx.xxx.xxx.xxy by rule: xxx.xxx.xxx.xxy-200
> TRACE 2017-02-07 18:08:56.021 [UPNPHelper] Receiving a M-SEARCH from [xxx.xxx.xxx.xxy:44473]: M-SEARCH * HTTP/1.1
> HOST: 239.255.255.250:1900
> MAN: "ssdp:discover"
> MX: 10
> ST: urn:schemas-upnp-org:device:MediaServer:1
> 
> 
> TRACE 2017-02-07 18:08:56.022 [UPNPHelper] Sending discovery [xxx.xxx.xxx.xxy:44473]: HTTP/1.1 200 OK<CRLF>

`CustomFFmpegOptions = -async 1 -fflags +genpts`

> ERROR 2017-02-07 18:09:36.177 [New I/O worker #8] There is no inputstream to return for test.mp4 [tsMuxeR] {Audio: aac (lc) (mp4a / 0x6134706d)}

@onon765trb onion When you say "_It isn't impossible to playback movie wuth tsmuxer_", do you really mean some MP4 movies only ?

**Comment by onon765trb** [comment]
@Nadahar Sorry for my bad explanation. What i say is that there is a case that ` -c:v copy` don' work without `-fflags +genpts`.

@Sami32 I tested only mp4 format. All mp4 format I  tested can't be playbacked. For other format, I can say nothing. I will try other formats when I  has some spare time.

**Comment by onon765trb** [comment]
Some causes of tsmuxer not well working may be clarified.

Summarized: good results are gotten by directly sending a input file to tsmuxer or a video stream sent to tsmuxer is muxed to mkv.

In the below tests, tsMuxeR-New and tsmuxer CUI  and GUI v 2.6.11 are used.

Firstly, the crash of tsmuxer
```
"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR"
Segmentation fault: 11
```


Secondly, getting infometaion tsmuxer. Profile and  Resolution are strange. a meta file and full log are attached.
```
DEBUG 2017-07-01 18:13:37.128 [HTTPv2 Request Worker 14] Starting transcode/remux of 1080.mkv with media info: Container: MATROSKA, Size: 5051075219, Video Bitrate: 6110208, Video Tracks: 1, Video Codec: h264 (main), Duration: 1:52:51, Video Resolution: 1920 x 1080, Frame Rate: 29.97, Audio Tracks: 1 [Audio Codec: aac (lc), Bitrate: 0, Channels: 2, Sample Frequency: 44100 Hz], Mime Type: video/x-matroska

DEBUG 2017-07-01 18:13:38.359 [Thread-46] Decoding H264 stream (track 1): Profile: Baseline@19.6  Resolution: 16:14p  Frame rate: not found
```
My manual test with tsmuer. using `-f rawvideo`, info can't be gotten.
```
/sw/bin/ffmpeg -i /Users/x/dlna/1080.mkv -vframes 1 -c:v copy -f h264 -bsf h264_mp4toannexb x_h264
"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR-new" "/Users/x/dlna/x_h264"
Network Optix tsMuxeR.  Version 2.2.3(b). www.networkoptix.com
Stream type: H.264
Stream ID:   V_MPEG4/ISO/AVC
Stream info: Profile: Main@4.0  Resolution: 1920:1080p  Frame rate: 29.97
Stream lang: 

/sw/bin/ffmpeg -ss 0 -i /Users/x/dlna/1080.mkv -c copy -f rawvideo -y x_rawvideo
"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR-new" /Users/x/dlna/x_rawvideo 
Network Optix tsMuxeR.  Version 2.2.3(b). www.networkoptix.com
Can't detect stream type
```

Thirdly , the matter of demuxer of tsmuxer CUI. tsmuxer may not address raw file (foo.h264) well.
```
i="/Users/x/dlna/1080.mkv"

file_name="${i##*/}"
file_name_without_ext="${file_name%.*}"
directory="${i%/*}"

cd "${directory}"
ffmpeg -i "$i" -c:v copy -an /Users/x/dlna/x_v.h264 -vn -c:a copy /Users/x/dlna/x_a.aac
echo "MUXOPT --no-pcr-on-video-pid --new-audio-pes --vbr  --vbv-len=500
V_MPEG4/ISO/AVC, /Users/x/dlna/x_v.h264, insertSEI, contSPS, track=1, lang=und
A_AAC, /Users/x/dlna/x_a.aac, track=2, lang=und" >"${file_name_without_ext}".meta

"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR-new" "${file_name_without_ext}".meta "${file_name_without_ext}".ts
Network Optix tsMuxeR.  Version 2.2.3(b). www.networkoptix.com
95.7% complete
Processed 0 video frames
Decoding AAC stream (track 2): Sample Rate: 44KHz  Channels: 2
100.0% complete
Flushing write buffer
Mux successful complete
Muxing time: 35 sec
```
In this test, the video stream isn't muxed. Only the audio stream is muxed. Even with tsmuxer CUI v 2.6.11, a result is the same.
With tsmuxer GUI v 2.6.11. this problem don't appear.

Additional tests gotten good results. A video stream is muxed to mkv and a audio stream is dump to a raw file. 
```
i="/Users/x/dlna/1080.mkv"

file_name="${i##*/}"
file_name_without_ext="${file_name%.*}"
directory="${i%/*}"

cd "${directory}"
ffmpeg -i "$i" -c:v copy -an /Users/x/dlna/x_v.mkv -vn -c:a copy /Users/x/dlna/x_a.aac
echo "MUXOPT --no-pcr-on-video-pid --new-audio-pes --vbr  --vbv-len=500
V_MPEG4/ISO/AVC, /Users/x/dlna/x_v.mkv, insertSEI, contSPS, track=1, lang=und
A_AAC, /Users/x/dlna/x_a.aac, track=2, lang=und" >"${file_name_without_ext}".meta

"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR-new" "${file_name_without_ext}".meta "${file_name_without_ext}".ts
Network Optix tsMuxeR.  Version 2.2.3(b). www.networkoptix.com
Decoding H264 stream (track 1): Profile: Main@4.0  Resolution: 1920:1080p  Frame rate: 29.97
H.264 muxing fps is not set. Get fps from stream. Value: 29.97
0.0% complete
Decoding AAC stream (track 2): Sample Rate: 44KHz  Channels: 2
99.7% complete
B-pyramid level 1 detected. Shift DTS to 2 frames
Processed 202935 video frames
100.0% complete
Flushing write buffer
Mux successful complete
Muxing time: 1 min 19 sec

echo "MUXOPT --no-pcr-on-video-pid --new-audio-pes --vbr  --vbv-len=500
V_MPEG4/ISO/AVC, "${i}", insertSEI, contSPS, track=1, lang=und
A_AAC, "${i}", track=2, lang=und" >"${file_name_without_ext}".meta

"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR-new" "${file_name_without_ext}".meta "${file_name_without_ext}".ts
Network Optix tsMuxeR.  Version 2.2.3(b). www.networkoptix.com
Decoding H264 stream (track 1): Profile: Main@4.0  Resolution: 1920:1080p  Frame rate: 29.97
H.264 muxing fps is not set. Get fps from stream. Value: 29.97
0.0% complete
Decoding AAC stream (track 2): Sample Rate: 44KHz  Channels: 2
99.5% complete
B-pyramid level 1 detected. Shift DTS to 2 frames
Processed 202935 video frames
100.0% complete
Flushing write buffer
Mux successful complete
Muxing time: 1 min 12 sec
```

Additional tests gotten bad results.  If a audio stream is muxed to mka. Its movie file can't be playbacked by mpv. Quick Time and Quick look can playback. But it take long time until starting and a audio stream isn't muxed. With tsmuxer GUI v 2.6.11. this problem don't appear.
```
i="/Users/x/dlna/1080.mkv"

file_name="${i##*/}"
file_name_without_ext="${file_name%.*}"
directory="${i%/*}"

cd "${directory}"
ffmpeg -i "$i" -c:v copy -an /Users/x/dlna/x_v.mkv -vn -c:a copy /Users/x/dlna/x_a.mka
echo "MUXOPT --no-pcr-on-video-pid --new-audio-pes --vbr  --vbv-len=500
V_MPEG4/ISO/AVC, /Users/x/dlna/x_v.mkv, insertSEI, contSPS, track=1, lang=und
A_AAC, /Users/x/dlna/x_a.mka, track=2, lang=und" >"${file_name_without_ext}".meta

"/Applications/Universal Media Server.app/Contents/Resources/osx/tsMuxeR-new" "${file_name_without_ext}".meta "${file_name_without_ext}".ts
Network Optix tsMuxeR.  Version 2.2.3(b). www.networkoptix.com
Decoding H264 stream (track 1): Profile: Main@4.0  Resolution: 1920:1080p  Frame rate: 29.97
H.264 muxing fps is not set. Get fps from stream. Value: 29.97
99.6% complete
B-pyramid level 1 detected. Shift DTS to 2 frames
Processed 202935 video frames
100.0% complete
Flushing write buffer
Mux successful complete
Muxing time: 1 min 15 sec
```

**Comment by SubJunk** [comment]
tsMuxeR has been disabled by default for years and is no longer being developed. We found that all versions have bugs that affect some users so it's best to just leave it disabled.

Please feel free to submit a pull request with code changes to improve our tsMuxeR support, but I won't be working on it personally

**Comment by onon765trb** [comment]
Thank you for your comment and work.
I can understand you because tsmuxer isn't developed and its source isn't opened.
My request is that not removing tsmuxer from ums as long as the problem of security don't arises.

**Comment by SubJunk** [comment]
Thanks for your understanding. I have no plans to remove tsMuxeR

**Comment by SubJunk** [comment]
I have reproduced this on macOS and it seems to be a problem with the tsMuxeR binary. Will fix
